fuzz/fuzz_targets/ruff_formatter_validity.rs

```diff
@@ -0,0 +1,103 @@
+//! Fuzzer harness which actively tries to find testcases that cause Ruff to introduce errors into
+//! the resulting file.
+
+#![no_main]
+
+use std::collections::HashMap;
+use std::sync::OnceLock;
+
+use libfuzzer_sys::{fuzz_target, Corpus};
+use ruff_linter::linter::ParseSource;
+use ruff_linter::settings::flags::Noqa;
+use ruff_linter::settings::LinterSettings;
+use ruff_linter::source_kind::SourceKind;
+use ruff_python_ast::PySourceType;
+use ruff_python_formatter::{format_module_source, PyFormatOptions};
+use similar::TextDiff;
+
+static SETTINGS: OnceLock<LinterSettings> = OnceLock::new();
+
+fn do_fuzz(case: &[u8]) -> Corpus {
+    // throw away inputs which aren't utf-8
+    let Ok(code) = std::str::from_utf8(case) else {
+        return Corpus::Reject;
+    };
+
+    // the settings are immutable to test_snippet, so we avoid re-initialising here
+    let linter_settings = SETTINGS.get_or_init(LinterSettings::default);
```

Comment:
Should we test with all rules (rather than just the default rule set)? 

If so, we then need to exclude the rules that we know are conflicting with the formatter

https://github.com/astral-sh/ruff/blob/da8a3af52409a5cd5034060a482ae93ac6abc27f/crates/ruff_cli/src/commands/format.rs#L735-L746

Reply:
Can we make this a public array so that we can remove them in the fuzzer? Then, they can be updated in the crate according to when the rules are eventually made compatible, if ever.

Reply:
We probably could. Let's go with the standard rules for now. Is it possible to parametrize fuzzers?

Reply:
@MichaReiser Kind of, but fuzzers should generally avoid it. Since we are already parameterised over input, parameterising over something else means we basically square our input space (which is going to make the fuzzer ineffective).